### PR TITLE
Refactored the Document Errors

### PR DESCRIPTION
**Pull request by MehrdadMzfr**
-Change the MSDataError to inherit from the NSError
-Added contractor to pass an inner error in order to pass/keep the error context

Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests?
* [x] Did you check UI tests on the sample app? They are not executed on CI.
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

-Change the MSDataError to inherit from the NSError
-Added contractor to pass an inner error in order to pass/keep the error context

### CODE REVIEW

File: AppCenterData/AppCenterData/Internal/Client/MSDataOperationProxy.m [hunk 1]
```diff
@@ -150,7 +152,8 @@ - (void)performOperation:(NSString *_Nullable)operation
                                          code:MSACDataErrorJSONSerializationFailed
                                      userInfo:@{NSLocalizedDescriptionKey : @"Dictionary contains values that cannot be serialized."}];
           MSLogError([MSData logTag], @"Error serializing document for local storage: %@", [jsonError localizedDescription]);
-          completionHandler([[MSDocumentWrapper alloc] initWithError:jsonError documentId:documentId]);
+          MSDataError *dataError = [[MSDataError alloc] initWithInnerError:jsonError code:0 message:nil];
```
Comment (clpolet) [inline AppCenterData/AppCenterData/Internal/Client/MSDataOperationProxy.m]: Are we sure about the code:0 message:nil?

File: AppCenterData/AppCenterData/Internal/Util/MSDocumentUtils.m [hunk 2]
```diff
@@ -64,7 +64,9 @@ + (MSDocumentWrapper *)documentWrapperFromData:(nullable NSData *)data
                                      userInfo:@{NSLocalizedDescriptionKey : @"Can't deserialize JSON payload"}];
     }
     MSLogError([MSData logTag], @"Error deserializing data: %@", [error description]);
-    return [[MSDocumentWrapper alloc] initWithError:error documentId:nil];
+
```
Comment (clpolet) [inline AppCenterData/AppCenterData/Internal/Util/MSDocumentUtils.m]: Remove extra line here.

File: AppCenterData/AppCenterData/Internal/Util/MSDocumentUtils.m [hunk 3]
```diff
@@ -185,11 +186,11 @@ + (MSDocumentWrapper *)documentWrapperFromDictionary:(NSDictionary *)dictionary
 
   // Validate dictionary
   if (![NSJSONSerialization isValidJSONObject:dictionary]) {
-    error = [[NSError alloc] initWithDomain:kMSACDataErrorDomain
-                                       code:MSACDataErrorJSONSerializationFailed
-                                   userInfo:@{NSLocalizedDescriptionKey : @"Dictionary contains values that cannot be serialized."}];
-    MSLogError([MSData logTag], @"Error deserializing data: %@", [error localizedDescription]);
-    return [[MSDocumentWrapper alloc] initWithError:error documentId:documentId];
+
```
Comment (clpolet) [inline AppCenterData/AppCenterData/Internal/Util/MSDocumentUtils.m]: Remove extra line.

File: AppCenterData/AppCenterData/Internal/Util/MSDocumentUtils.m [hunk 4]
```diff
@@ -198,24 +199,23 @@ + (MSDocumentWrapper *)documentWrapperFromDictionary:(NSDictionary *)dictionary
   }
 
   // Deserialize document.
+  MSDataError *dataError = nil;
   id<MSSerializableDocument> deserializedValue;
   if (!error && ![MSDocumentUtils isReferenceDictionaryWithKey:dictionary key:kMSDocumentKey keyType:[NSDictionary class]]) {
-    error = [[NSError alloc] initWithDomain:kMSACDataErrorDomain
-                                       code:MSACDataErrorJSONSerializationFailed
-                                   userInfo:@{NSLocalizedDescriptionKey : @"Can't deserialize document (missing document property)"}];
+
```
Comment (clpolet) [inline AppCenterData/AppCenterData/Internal/Util/MSDocumentUtils.m]: Remove extra line here.

File: AppCenterData/AppCenterData/MSData.m [hunk 5]
```diff
@@ -407,17 +405,17 @@ - (void)listDocumentsWithType:(Class)documentType
                                   id jsonPayload = [NSJSONSerialization JSONObjectWithData:(NSData *)data
                                                                                    options:0
                                                                                      error:&deserializeError];
+
```
Comment (clpolet) [inline AppCenterData/AppCenterData/MSData.m]: Remove extra line here.

File: AppCenterData/AppCenterData/Internal/Util/MSDocumentUtils.m [hunk 6]
```diff
@@ -121,12 +124,10 @@ + (MSDocumentWrapper *)documentWrapperFromDictionary:(NSObject *)object
       ![MSDocumentUtils isReferenceDictionaryWithKey:object key:kMSPartitionKey keyType:[NSString class]]) {
 
     // Prepare and return error.
-    NSError *error = [[NSError alloc]
-        initWithDomain:kMSACDataErrorDomain
-                  code:MSACDataErrorJSONSerializationFailed
-              userInfo:@{NSLocalizedDescriptionKey : @"Can't deserialize document (missing system properties or partition key)"}];
-    MSLogError([MSData logTag], @"Error deserializing data: %@", [error localizedDescription]);
-    return [[MSDocumentWrapper alloc] initWithError:error documentId:nil];
+    NSString *errorMessage = @"Can't deserialize document (missing system properties or partition key)";
```
Comment (jaelim-ms) [inline AppCenterData/AppCenterData/Internal/Util/MSDocumentUtils.m]: Period (`.`) at the end of the log message.

File: AppCenterData/AppCenterData/MSData.m [hunk 7]
```diff
@@ -520,23 +519,27 @@ - (void)upsertFromCosmosDbWithPartition:(NSString *)partition
...
                                                            partition:partition
                                                             document:[document serializeToDictionary]];
   if (![NSJSONSerialization isValidJSONObject:dic]) {
-    serializationError =
-        [[NSError alloc] initWithDomain:kMSACDataErrorDomain
-                                   code:MSACDataErrorJSONSerializationFailed
-                               userInfo:@{NSLocalizedDescriptionKey : @"Document dictionary contains values that cannot be serialized."}];
-    MSLogError([MSData logTag], @"Error serializing data: %@", [serializationError localizedDescription]);
-    completionHandler([[MSDocumentWrapper alloc] initWithError:serializationError documentId:documentId]);
+
```
Comment (clpolet) [inline AppCenterData/AppCenterData/MSData.m]: Remove extra lline

File: AppCenterData/AppCenterData/MSDataError.h [hunk 8]
```diff
@@ -3,34 +3,38 @@
...
  *
- * @param error An error object.
+ * @param userInfo Contains context about the actual error.
+ * @param code Error code.
  *
- * @return A new `MSDataError` instance.
+ * @return Instance of error object.
  */
-- (instancetype)initWithError:(NSError *)error; // TODO move to internal
+- (instancetype)initWithUserInfo:(NSDictionary *)userInfo code:(NSInteger)code;
```
Comment (jaelim-ms) [inline AppCenterData/AppCenterData/MSDataError.h]: `- (instancetype)initWithErrorCode:(NSInteger)errorCode userInfo(NSDictionary *)userInfo;`

File: AppCenterData/AppCenterData/MSDataError.h [hunk 9]
```diff
@@ -3,34 +3,38 @@
...
+ * Create a MSDataError object.
+ *
+ * @param innerError Inner error that provides more context about the actual error.
+ * @param code Error code.
+ * @param message Error message.
+ *
+ * @return Instance of error object.
  */
-@property(nonatomic, readonly) NSInteger errorCode;
+- (instancetype)initWithInnerError:(NSError *_Nullable)innerError code:(NSInteger)code message:(NSString *_Nullable)message;
```
Comment (jaelim-ms) [inline AppCenterData/AppCenterData/MSDataError.h]:
```
- (instancetype)initWithErrorCode:(NSInteger)errorCode innerError:(NSError *_Nullable)innerError message:(NSString *_Nullable)message;
```

**Pull request merged**